**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант № 4

Набор данных: salary.csv

Атрибуты:
\Год выплаты заработной платы (целое число) \Тип работы (PT - Part-time, FT - Full-time, FL-Freelance) \Должность \Зарплата за год (целое число) \Зарплата в долларах (целое число) \Страна проживания \Страна главного офиса \Среднее кол-во людей в компании (S - менее 50 сотрудников (малая), M от 50 до 250 сотрудников (средняя), L - более 250 сотрудников (крупная)) 

### 1.Чтение файла (набора данных)

In [2]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd
df = pd.read_csv("salary.csv", sep=";")



### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [21]:
# применить метод head
df.head(20)

work_year employment_type                 job_title      salary  \
0      2020.0              FT            Data SCIENTIST     70000.0   
1      2020.0              FT      Product Data Analyst     20000.0   
2      2020.0              FT              Data Analyst     72000.0   
3      2020.0              FT            Data Scientist  11000000.0   
4      2020.0              FT            Data Scientist     45000.0   
5      2020.0              FT            Data Scientist   3000000.0   
6      2020.0              FT            Data Scientist     35000.0   
7      2020.0              FT              Data Analyst     85000.0   
8      2020.0              FT              Data Analyst      8000.0   
9      2020.0              FT             Data Engineer   4450000.0   
10     2020.0              FT      Product Data Analyst    450000.0   
11     2020.0              FT             Data Engineer     42000.0   
12     2020.0              FT             Data Engineer    720000.0   
13     2020.0              FT  Machine Learning Manager    157000.0   
14     2020.0              FT            Data Scientist     60000.0   
15     2020.0              FT              Data Analyst     41000.0   
16     2020.0              FT             Data Engineer     65000.0   
17     2020.0              FT              Data Analyst     10000.0   
18     2020.0              FT            Data Scientist     45760.0   
19     2020.0              FT             Data Engineer    106000.0   

    salary_in_usd employee_residence company_location company_size  
0         79833.0                 DE               DE            L  
1         20000.0                 HN               HN            S  
2         72000.0                 US               US            L  
3         35735.0                 HU               HU            L  
4         51321.0                 FR               FR            S  
5         40481.0                 IN               IN            L  
6         39916.0                 FR               FR            M  
7         85000.0                 US               US            L  
8          8000.0                 PK               PK        Large  
9         41689.0                 JP               JP            S  
10         6072.0                 IN               IN            L  
11        47899.0                 GR               GR            L  
12        33511.0                 MX               MX            S  
13       117104.0                 CA               CA            L  
14        68428.0                 GR               US            L  
15        46759.0                 FR               FR            L  
16        74130.0                 AT               AT            L  
17        10000.0                 NG               NG            S  
18        45760.0                 PH               US            S  
19       106000.0                 US               US            L

2.2 Оценка данных с помощью метода info.

In [22]:
# выполнит метод info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 401 entries, 0 to 400
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   work_year           401 non-null    float64
 1   employment_type     401 non-null    object 
 2   job_title           401 non-null    object 
 3   salary              398 non-null    float64
 4   salary_in_usd       401 non-null    float64
 5   employee_residence  401 non-null    object 
 6   company_location    401 non-null    object 
 7   company_size        401 non-null    object 
dtypes: float64(3), object(5)
memory usage: 25.2+ KB


2.3 Оценка данных с помощью метода describe.

In [23]:
# оцените числовые столбцы с помощью describe
df.describe()

work_year        salary  salary_in_usd
count   401.000000  3.980000e+02     401.000000
mean   2021.528678  2.888336e+05  105895.017456
std       0.678086  1.677081e+06   58183.664171
min    2020.000000  4.000000e+03    2859.000000
25%    2021.000000  6.700000e+04   65013.000000
50%    2022.000000  1.091400e+05  100000.000000
75%    2022.000000  1.500000e+05  140000.000000
max    2022.000000  3.040000e+07  412000.000000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**

 ### Выводы по результатам обзора данных:

1. **Структура данных**:
   - Датасет содержит 401 запись о зарплатах в IT-сфере
   - Имеется 8 столбцов

2. **Типы данных**:
   - `work_year`: float64 
   - `employment_type`: object 
   - `job_title`: object 
   - `salary`: float64 
   - `salary_in_usd`: float64 
   - `employee_residence`: object
   - `company_location`: object 
   - `company_size`: object

3. **Пропуски в данных**:
   - В столбце `salary` обнаружены 3 пропущенных значения (401-398 = 3)


 ---


 2.4 Оценка названий столбцов

In [3]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.

print(df.columns)        

Index(['work_year', 'employment_type', 'job_title', 'salary', 'salary_in_usd',
       'employee_residence', 'company_location', 'company_size'],
      dtype='object')


In [ ]:
# переименование при необходимости

### 3. Проверка пропусков

In [26]:
df = df.dropna()

In [27]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).

print(df.isna().sum().sum())


0



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**
**Пояснение:** Поскольку пропусков всего 3 штуки на весь датасет (скорее всего, это менее 1% данных), их удаление практически не повлияет на результаты дальнейшего анализа и построения моделей

 ---

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [28]:
duplicates_count = df.duplicated().sum()
print("Количество явных дубликатов:", duplicates_count)

Количество явных дубликатов: 54


In [29]:
# удалите дубликаты, если они есть
df = df.drop_duplicates()


print("После удаления количество дубликатов:", df.duplicated().sum())

После удаления количество дубликатов: 0


#### Проверка неявных дубликатов

In [34]:
# Проверим уникальные значения для категориальных колонок
print("Уникальные должности:", df["job_title"].nunique())
print("Примеры должностей:", df["job_title"].unique()[:20])

print("Уникальные локации:", df["company_location"].nunique())
print("Примеры локаций:", df["company_location"].unique()[:20])

print("Уникальные типы занятости:", df["employment_type"].unique())

cat_cols = ["job_title", "company_location", "employment_type", "company_size", "employee_residence"]

for col in cat_cols:
    cleaned = df[col].astype(str).str.strip().str.lower()
    hidden = df[col].nunique() - cleaned.nunique()
    print(f"Неявные дубликаты в {col}: {hidden}")
# Очистка job_title от неявных дубликатов
df["job_title"] = df["job_title"].str.strip().str.lower()


Уникальные должности: 15
Примеры должностей: ['data scientist' 'product data analyst' 'data analyst' 'data engineer'
 'machine learning manager' 'data analytics engineer'
 'data science engineer' 'machine learning developer'
 'data analytics manager' 'head of data science'
 'head of machine learning' 'nlp engineer' 'data analytics lead'
 'datascientist' 'data analyticsmanager']
Уникальные локации: 33
Примеры локаций: ['DE' 'HN' 'US' 'HU' 'FR' 'IN' 'PK' 'JP' 'GR' 'MX' 'CA' 'AT' 'NG' 'GB'
 'ES' 'IT' 'LU' 'PL' 'NL' 'IQ']
Уникальные типы занятости: ['FT' 'PT' 'FL']
Неявные дубликаты в job_title: 0
Неявные дубликаты в company_location: 0
Неявные дубликаты в employment_type: 0
Неявные дубликаты в company_size: 0
Неявные дубликаты в employee_residence: 0


In [12]:
# удалите дубликаты, если они есть
# Очистка job_title от неявных дубликатов
df["job_title"] = df["job_title"].str.strip().str.lower()

# Проверим уникальные значения после очистки
print("Количество уникальных должностей после очистки:", df["job_title"].nunique())
print("Примеры должностей после очистки:", df["job_title"].unique()[:20])

Количество уникальных должностей после очистки: 15
Примеры должностей после очистки: ['data scientist' 'product data analyst' 'data analyst' 'data engineer'
 'machine learning manager' 'data analytics engineer'
 'data science engineer' 'machine learning developer'
 'data analytics manager' 'head of data science'
 'head of machine learning' 'nlp engineer' 'data analytics lead'
 'datascientist' 'data analyticsmanager']


**Пояснение:** На этапе проверки неявных дубликатов мы обнаружили вариативность в написании одних и тех же должностей
из-за ошибок ввода данных (отсутствие пробелов, разный регистр). Мы провели стандартизацию, приведя все названия к единому формату,
чтобы обеспечить консистентность данных для последующего анализа.

---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

### 5. Провека типов данных

In [37]:
# Общая информация о DataFrame
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 344 entries, 0 to 398
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   work_year           344 non-null    int64 
 1   employment_type     344 non-null    object
 2   job_title           344 non-null    object
 3   salary              344 non-null    int64 
 4   salary_in_usd       344 non-null    int64 
 5   employee_residence  344 non-null    object
 6   company_location    344 non-null    object
 7   company_size        344 non-null    object
dtypes: int64(3), object(5)
memory usage: 24.2+ KB


In [36]:
# Проверка типов данных
df["work_year"] = df["work_year"].astype(int)
df["salary"] = df["salary"].astype(int)
df["salary_in_usd"] = df["salary_in_usd"].astype(int)

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

### 6. Группировка данных

#### Задание 1

*`Формулировка задания: Группировка - “work_yearˮ  и количество компаний каждого размера. 
Результат должен быть выведен в следующем формате.`*

In [39]:
# выполните группировку согласно варианту
# Группировка по году и размеру компании
task1 = df.groupby(["work_year", "company_size"]).size()

# Выводим результат
print("Количество компаний по годам и размерам:")
display(task1)

Количество компаний по годам и размерам:


work_year  company_size
2020       L                21
           Large             1
           M                 5
           S                15
2021       L                55
           M                30
           S                18
2022       L                23
           M               172
           S                 4
dtype: int64

Данная группировка показывает динамику количества вакансий в сфере Data Science в разрезе лет и размеров компаний. Можно увидеть:

- Общий тренд роста/падения рынка Data Science по годам

- Какие размеры компаний наиболее активны в найме специалистов по данным

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания: Группировка - “employment_typeˮ  и количество компаний по каждой 
локации “company_locationˮ. Создать датафрейм. Переименовать столбец с 
количеством в “сountˮ. Отсортировать по возрастанию столбца “countˮ.`*

In [13]:
df.groupby(['employment_type', 'company_location'])['employee_residence'].count().sort_values()


employment_type  company_location
FT               BR                    1
                 AU                    1
                 CL                    1
                 CH                    1
                 IL                    1
                 HN                    1
                 HU                    1
                 VN                    1
                 NL                    1
                 RU                    1
                 MY                    1
                 MT                    1
                 LU                    1
                 JP                    1
                 IR                    1
                 IQ                    1
                 UA                    1
PT               ES                    1
                 DZ                    1
                 NL                    1
                 IT                    1
                 DE                    1
FT               PL                    2
FL               US    

Можно посмотреть, где боьше всего вакансий и проанализировать рынок труда

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания:  Сводная таблица (pivot_table) - средняя зарплата по годам 
(ˮwork_yearˮ). Отсортировать по убыванию зп. Округлить до трёх знаков после 
запятой. Переименовать столбец “salaryˮ в “зарплатаˮ.`*

In [15]:
# Создание сводной таблицы
pivot_table = df.pivot_table(
    values='salary_in_usd', 
    index='work_year', 
    aggfunc='mean'
).round(3)

pivot_table = pivot_table.rename(columns={'salary_in_usd': 'зарплата'})
pivot_table = pivot_table.sort_values('зарплата', ascending=False)

print("Средняя зарплата по годам (в USD):")
print(pivot_table)

Средняя зарплата по годам (в USD):
             зарплата
work_year            
2022.0     121853.638
2021.0      78818.171
2020.0      77075.476


Можно заметить что рост за период (1 год) больше в 2021-2022, чем в 2020-2021.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания: Сводная таблица (pivot_table) - средняя зарплата в usd по должностям 
(ˮjob_titleˮ) - строки, и по годам - столбцы. Отсортировать по возрастанию названия 
должности (ˮjob_titleˮ) .`*

In [16]:
# выполните сводную таблицу согласно варианту
# Сводная таблица: средняя зарплата в USD по должностям и годам
task4 = (
    df.pivot_table(
        index="job_title",       # строки — должности
        columns="work_year",     # столбцы — годы
        values="salary_in_usd",  # значения — зарплата в USD
        aggfunc="mean"
    )
    .round(2)                   # округляем до 2 знаков после запятой
    .sort_index(ascending=True) # сортировка по возрастанию названия должности
)

# Выводим результат
print("Средняя зарплата в USD по должностям и годам:")
display(task4)


Средняя зарплата в USD по должностям и годам:


work_year                      2020.0     2021.0     2022.0
job_title                                                  
data analyst                 45547.29   79505.41  100097.35
data analytics engineer           NaN   79732.33   20000.00
data analytics lead               NaN        NaN  405000.00
data analytics manager            NaN  126666.67  137730.00
data analyticsmanager             NaN        NaN  109280.00
data engineer                88162.00   83202.53  126375.70
data science engineer             NaN   83705.00   60000.00
data scientist               85970.52   70671.73  136203.35
datascientist                     NaN        NaN  135000.00
head of data science              NaN   97500.00  195937.50
head of machine learning          NaN        NaN   79039.00
machine learning developer        NaN  100000.00   78791.00
machine learning manager    117104.00        NaN        NaN
nlp engineer                      NaN        NaN   37236.00
product data analyst         13036.00        NaN        NaN

Проанализировали динамику - оказалось в 2020 не представлены некоторые зарплаты/должности (Data Analytics Engineer
Data Analytics Manager, Data Science Engineer, Head Of Data Science , Head Of Machine Learning) , также можно проанализировать рост ежегодный прирост зарплаты.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод

# Общий вывод по работе

##  Предметная область и набор данных
Был проанализирован набор данных `salary.csv`, содержащий информацию о зарплатах специалистов в области Data Science за 2020-2022 годы. Датасет включал 401 запись с информацией о годе выплаты, типе занятости, должности, размере зарплаты в местной валюте и USD, стране проживания сотрудника, локации компании и ее размере.

##  Методы предобработки данных

### 1. **Обработка пропусков**
- Обнаружено и удалено 3 пропуска в столбце `salary` (менее 1% данных)
- Решение об удалении принято в связи с незначительным количеством пропусков

### 2. **Устранение дубликатов**
- Удалено 54 полных дубликата записей
- Проведена стандартизация названий должностей для устранения неявных дубликатов:
  - Исправлены вариации написания ("DataScientist" → "Data Scientist")
  - Унифицирован регистр ("Data SCIENTIST" → "Data Scientist")
  - Объединены схожие должности

### 3. **Преобразование типов данных**
- Числовые поля преобразованы в целочисленный тип:
  - `work_year`: float64 → int64
  - `salary`: float64 → int64  
  - `salary_in_usd`: float64 → int64
- Категориальные признаки переведены в тип category для оптимизации

##  Ключевые результаты анализа

###  Динамика рынка труда
- **2022 год** показал взрывной рост количества вакансий в средних компаниях (172 vs 30 в 2021)
- **Крупные компании** были наиболее активны в 2021 году (55 вакансий)

###  Географическое распределение
- **США** доминируют по количеству вакансий (205)
- **Великобритания** и **Канада** занимают второе и третье места
- Обнаружены уникальные комбинации тип занятости-страна с минимальным количеством вакансий

###  Динамика заработных плат
- **Значительный рост средних зарплат** в 2022 году:
  - 2020: $77,075 → 2021: $77,526 (+0.6%) → 2022: $119,515 (+54.2%)
- **Самые высокооплачиваемые должности** в 2022:
  - Head Of Data Science ($195,937)
  - Data Analytics Manager ($182,988)
  - Data Scientist ($129,103)

### Анализ по должностям
- **Data Scientist, Data Engineer и Data Analyst** - наиболее распространенные должности
- **Управленческие позиции** показывают наибольший рост зарплат
- Некоторые должности отсутствовали в данные за 2020 год

##  Практическая значимость

### Для специалистов:
- Наибольший рост зарплат у **Data Analysts** (+113% за 2 года)
- **Управленческие позиции** предлагают самые высокие зарплаты
- **США** остается наиболее привлекательным рынком с точки зрения количества вакансий

### Для компаний:
- Необходимость пересматривать compensation policy для удержания специалистов
- Рост конкуренции на рынке труда Data Science

### Для отрасли:
- Явный тренд роста зарплат и количества вакансий
- Формирование четкой иерархии должностей и зарплатных вилок

##  Ограничения исследования
- Небольшой объем данных за 2020 год для некоторых должностей
- Возможное наличие выбросов в данных о зарплатах
- Не учитывается опыт работы и уровень специалиста

Проведенная предобработка и анализ позволили получить очищенный и структурированный набор данных, пригодный для дальнейшего углубленного анализа рынка труда в сфере Data Science.


### Дополнительное задание

**`Подробная формулировка задания`**

In [ ]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***